App crashes on startup on API 16

Caused by: java.lang.NullPointerException: CameraUpdateFactory is not initialized
 at cc.softwarefactory.lokki.android.fragments.MapViewFragment.onResume(MapViewFragment.java:201)

--- a/App/src/main/java/cc/softwarefactory/lokki/android/activities/MainActivity.java
+++ b/App/src/main/java/cc/softwarefactory/lokki/android/activities/MainActivity.java
@@ -433,7 +433,8 @@ public class MainActivity extends AppCompatActivity implements NavigationDrawerF
                     }
                 }
 
-                SearchView searchView=(SearchView) MenuItemCompat.getActionView(menu.findItem(R.id.search));
+                //Set up the search bar
+                final SearchView searchView=(SearchView) MenuItemCompat.getActionView(menu.findItem(R.id.search));
                 searchView.setQueryHint(getString(R.string.search_hint));
                 searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener(){
 
@@ -446,6 +447,12 @@ public class MainActivity extends AppCompatActivity implements NavigationDrawerF
                     @Override
                     public boolean onQueryTextSubmit(String query)
                     {
+                        //Removes focus from the search field in order to prevent multiple key events from
+                        //launching this callback. See:
+                        //http://stackoverflow.com/questions/17874951/searchview-onquerytextsubmit-runs-twice-while-i-pressed-once
+                        searchView.clearFocus();
+
+                        //Launch search activity
                         Intent intent= new Intent(mainactivity,SearchActivity.class);
                         Log.d(TAG,"Search Query submitted");
                         intent.putExtra(SearchActivity.QUERY_MESSAGE, query);

--- a/App/src/main/java/cc/softwarefactory/lokki/android/fragments/MapViewFragment.java
+++ b/App/src/main/java/cc/softwarefactory/lokki/android/fragments/MapViewFragment.java
@@ -140,6 +140,10 @@ public class MapViewFragment extends Fragment {
 
     //store current map state on SharedPreferences
     public void storeMapState(){
+        if (map == null){
+            Log.w(TAG, "No map, can't save current location");
+            return;
+        }
         Double lat = map.getCameraPosition().target.latitude;
         Double lon =  map.getCameraPosition().target.longitude;
 
@@ -194,7 +198,7 @@ public class MapViewFragment extends Fragment {
         AnalyticsUtils.screenHit(getString(R.string.analytics_screen_map));
 
 
-        if(MainApplication.emailBeingTracked == null){
+        if(MainApplication.emailBeingTracked == null && map != null){
             if(startLocation == null){
               loadMapState();
             }
@@ -246,6 +250,7 @@ public class MapViewFragment extends Fragment {
         map = fragment.getMap();
 
         if (map == null) {
+            Log.e(TAG, "Could not create map!");
             return;
         }
 
@@ -426,6 +431,10 @@ public class MapViewFragment extends Fragment {
     private BroadcastReceiver goToReceiver = new BroadcastReceiver() {
         @Override
         public void onReceive(Context context, Intent intent) {
+            if (map == null){
+                Log.w(TAG, "null map, not moving camera");
+                return;
+            }
             Log.d(TAG, "goToReceiver onReceive");
             //Parse coordinates from extra data
             String coords = intent.getStringExtra(GO_TO_COORDS);

--- a/App/src/main/java/cc/softwarefactory/lokki/android/services/LocationService.java
+++ b/App/src/main/java/cc/softwarefactory/lokki/android/services/LocationService.java
@@ -415,7 +415,7 @@ public class LocationService extends Service implements LocationListener, Google
     @Override
     public void onDestroy() {
         Log.d(TAG, "onDestroy called");
-        if(wakeLock.isHeld()) {
+        if(wakeLock != null && wakeLock.isHeld()) {
             wakeLock.release();
         }
         stopForeground(true);

